File: packages/backend/api/wallet/controllers/wallet.js
'use strict';

const walletService = require('../services/wallet');
const utils = require('../services/utils');
const { sendPushNotification } = require('../../helpers');

/**
 * Wallet Controller
 * User-facing wallet endpoints
 */

module.exports = {
  /**
   * GET /api/wallet/balance
   * Get current user's wallet balance
   */
  async getBalance(ctx) {
    try {
      const userId = ctx.state.user.id;

      if (!userId) {
        return ctx.unauthorized('Authentication required');
      }

      const balance = await walletService.getBalance(userId);
      ctx.send(utils.successResponse(balance));
    } catch (error) {
      strapi.log.error('[Wallet] getBalance error:', error);
      ctx.badRequest(utils.errorResponse('WALLET_ERROR', error.message));
    }
  },

  /**
   * GET /api/wallet/transactions
   * Get user's wallet transaction history
   */
  async getTransactions(ctx) {
    try {
      const userId = ctx.state.user.id;

      if (!userId) {
        return ctx.unauthorized('Authentication required');
      }

      const { limit, offset, type, status, fromDate, toDate } = ctx.query;

      const result = await walletService.getTransactions({
        userId,
        limit: limit || 20,
        offset: offset || 0,
        type,
        status,
        fromDate,
        toDate,
      });

      ctx.send(utils.successResponse(result));
    } catch (error) {
      strapi.log.error('[Wallet] getTransactions error:', error);
      ctx.badRequest(utils.errorResponse('WALLET_ERROR', error.message));
    }
  },

  /**
   * POST /api/wallet/top-up
   * Initiate wallet top-up transaction
   * Note: This is a placeholder. Real implementation needs Omise integration
   */
  async topUp(ctx) {
    try {
      const userId = ctx.state.user.id;

      if (!userId) {
        return ctx.unauthorized('Authentication required');
      }

      const { amount, paymentMethod, returnUrl } = ctx.request.body;

      // Validate amount
      const validation = utils.validateAmount(amount, 'top_up');
      if (!validation.valid) {
        return ctx.badRequest(utils.errorResponse('WALLET_004', validation.error));
      }

      // Calculate fee
      const fee = utils.calculateFee(amount, paymentMethod);
      const netAmount = amount - fee;

      // Get request metadata
      const metadata = utils.getRequestMetadata(ctx);

      // TODO: Integrate with Omise payment gateway
      // For now, create a pending transaction
      const transactionId = await utils.generateTransactionId();

      // This is a simplified version. In production, you would:
      // 1. Create Omise charge
      // 2. Get payment URL/QR code
      // 3. Create pending transaction
      // 4. Handle webhook for payment confirmation
      // 5. Update transaction status to completed

      ctx.send(utils.successResponse({
        transactionId,
        amount,
        fee,
        netAmount,
        paymentMethod,
        status: 'pending',
        // In production, include Omise charge data here
        message: 'Payment gateway integration required',
        createdAt: new Date(),
      }));
    } catch (error) {
      strapi.log.error('[Wallet] topUp error:', error);
      ctx.badRequest(utils.errorResponse('WALLET_005', error.message));
    }
  },

  /**
   * POST /api/wallet/pay
   * Pay for order using wallet balance
   */
  async pay(ctx) {
    try {
      const userId = ctx.state.user.id;

      if (!userId) {
        return ctx.unauthorized('Authentication required');
      }

      const {
        amount,
        referenceType,
        referenceId,
        branchId,
        description,
        usePoints = 0,
      } = ctx.request.body;

      // Validate amount
      const validation = utils.validateAmount(amount, 'payment');
      if (!validation.valid) {
        return ctx.badRequest(utils.errorResponse('WALLET_004', validation.error));
      }

      // Get request metadata
      const metadata = utils.getRequestMetadata(ctx);

      // Create payment transaction
      const transaction = await walletService.createTransaction({
        userId,
        type: 'payment',
        amount,
        referenceType,
        referenceId,
        description: description || `Payment for ${referenceType} ${referenceId}`,
        metadata: {
          branchId,
          usePoints,
        },
        ...metadata,
      });

      // TODO: Handle points usage if usePoints > 0
      // TODO: Calculate points earned from this payment

      const pointsEarned = Math.floor(amount * 0.1); // 10% points earning rate

      // Send push notification for payment
      try {
        await sendPushNotification({
          content: `Payment of ฿${amount.toFixed(2)} successful${pointsEarned > 0 ? ` (+${pointsEarned} points)` : ''}`,
          heading: 'Payment Successful',
          external_ids: [userId.toString()],
          additionalData: {
            type: 'wallet_payment',
            transactionId: transaction.id,
            amount,
            referenceType,
            referenceId,
            pointsUsed: usePoints,
            pointsEarned,
            balanceAfter: parseFloat(transaction.balance_after),
            timestamp: new Date().toISOString(),
          },
        });

        strapi.log.info('[Wallet] Push notification sent for payment:', transaction.id);
      } catch (notificationError) {
        // Don't fail the payment if notification fails
        strapi.log.error('[Wallet] Failed to send push notification:', notificationError);
      }

      ctx.send(utils.successResponse({
        transactionId: transaction.id,
        type: 'payment',
        amount: parseFloat(transaction.amount),
        pointsUsed: usePoints,
        pointsEarned,
        balanceBefore: parseFloat(transaction.balance_before),
        balanceAfter: parseFloat(transaction.balance_after),
        status: transaction.status,
        referenceId,
        completedAt: transaction.completed_at,
      }));
    } catch (error) {
      strapi.log.error('[Wallet] pay error:', error);

      if (error.message === 'WALLET_001') {
        return ctx.badRequest(utils.errorResponse('WALLET_001', 'Insufficient balance'));
      }
      if (error.message === 'WALLET_002') {
        return ctx.badRequest(utils.errorResponse('WALLET_002', 'Wallet frozen'));
      }

      ctx.badRequest(utils.errorResponse('WALLET_ERROR', error.message));
    }
  },

  /**
   * POST /api/wallet/voucher/redeem
   * Redeem voucher code for wallet credit
   */
  async redeemVoucher(ctx) {
    try {
      const userId = ctx.state.user.id;

      if (!userId) {
        return ctx.unauthorized('Authentication required');
      }

      const { code } = ctx.request.body;

      if (!code) {
        return ctx.badRequest(utils.errorResponse('WALLET_ERROR', 'Voucher code required'));
      }

      // Get request metadata
      const metadata = utils.getRequestMetadata(ctx);

      // Redeem voucher
      const transaction = await walletService.redeemVoucher({
        userId,
        code: code.toUpperCase(),
        ...metadata,
      });

      ctx.send(utils.successResponse({
        transactionId: transaction.id,
        voucherCode: code,
        amount: parseFloat(transaction.amount),
        balanceBefore: parseFloat(transaction.balance_before),
        balanceAfter: parseFloat(transaction.balance_after),
        description: transaction.description,
        createdAt: transaction.created_at,
      }));
    } catch (error) {
      strapi.log.error('[Wallet] redeemVoucher error:', error);

      if (error.message === 'WALLET_007') {
        return ctx.badRequest(utils.errorResponse('WALLET_007', 'Voucher expired or invalid'));
      }
      if (error.message === 'WALLET_008') {
        return ctx.badRequest(utils.errorResponse('WALLET_008', 'Voucher limit reached'));
      }

      ctx.badRequest(utils.errorResponse('WALLET_ERROR', error.message));
    }
  },

  /**
   * POST /api/wallet/points/convert
   * Convert loyalty points to wallet credit
   */
  async convertPoints(ctx) {
    try {
      const userId = ctx.state.user.id;

      if (!userId) {
        return ctx.unauthorized('Authentication required');
      }

      const { points } = ctx.request.body;

      if (!points || points <= 0) {
        return ctx.badRequest(utils.errorResponse('WALLET_ERROR', 'Invalid points amount'));
      }

      // Conversion rate: 10 points = 1 THB (configurable)
      const conversionRate = 10;
      const walletCredit = points / conversionRate;

      // Check if user has enough points
      const userPoints = await strapi.connections.default.raw(`
        SELECT SUM(point) as total_points
        FROM point_log
        WHERE user_id = ?
      `, [userId]);

      const totalPoints = userPoints[0][0]?.total_points || 0;

      if (totalPoints < points) {
        return ctx.badRequest(utils.errorResponse('WALLET_ERROR', 'Insufficient points'));
      }

      // Get request metadata
      const metadata = utils.getRequestMetadata(ctx);

      // Create conversion transaction
      const transaction = await walletService.createTransaction({
        userId,
        type: 'conversion',
        amount: walletCredit,
        description: `Converted ${points} points to wallet credit`,
        metadata: {
          pointsConverted: points,
          conversionRate,
        },
        ...metadata,
      });

      // Deduct points from user
      await strapi.connections.default.raw(`
        INSERT INTO point_log (user_id, point, description, created_at)
        VALUES (?, ?, ?, NOW())
      `, [userId, -points, `Converted to wallet credit (${walletCredit} THB)`]);

      const pointsRemaining = totalPoints - points;

      // Send push notification for points conversion
      try {
        await sendPushNotification({
          content: `${points} points converted to ฿${walletCredit.toFixed(2)} wallet credit`,
          heading: 'Points Redeemed',
          external_ids: [userId.toString()],
          additionalData: {
            type: 'wallet_points_conversion',
            transactionId: transaction.id,
            pointsConverted: points,
            conversionRate,
            walletCredit,
            pointsRemaining,
            balanceAfter: parseFloat(transaction.balance_after),
            timestamp: new Date().toISOString(),
          },
        });

        strapi.log.info('[Wallet] Push notification sent for points conversion:', transaction.id);
      } catch (notificationError) {
        // Don't fail the conversion if notification fails
        strapi.log.error('[Wallet] Failed to send push notification:', notificationError);
      }

      ctx.send(utils.successResponse({
        transactionId: transaction.id,
        pointsConverted: points,
        conversionRate,
        walletCredit,
        pointsRemaining,
        balanceBefore: parseFloat(transaction.balance_before),
        balanceAfter: parseFloat(transaction.balance_after),
        createdAt: transaction.created_at,
      }));
    } catch (error) {
      strapi.log.error('[Wallet] convertPoints error:', error);
      ctx.badRequest(utils.errorResponse('WALLET_ERROR', error.message));
    }
  },

  /**
   * GET /api/wallet/settings
   * Get public wallet settings (for application use)
   * Returns only public-facing settings like fees, limits, and conversion rates
   */
  async getSettings(ctx) {
    try {
      // Get settings from database
      const settingsRows = await strapi.connections.default.raw(`
        SELECT setting_key, setting_value
        FROM wallet_transfer_settings
      `);

      const settings = {};
      if (settingsRows && settingsRows[0]) {
        settingsRows[0].forEach(row => {
          settings[row.setting_key] = row.setting_value;
        });
      }

      // Return public-facing settings
      ctx.send(utils.successResponse({
        pointConversion: {
          rate: parseFloat(settings.point_conversion_rate || 1),
          minimumRedemption: parseInt(settings.point_min_redemption || 100),
          requiresApproval: settings.point_redemption_requires_approval === 'true',
        },
        transfer: {
          feePercentage: parseFloat(settings.transfer_fee_percentage || 0),
          feeFixed: parseFloat(settings.transfer_fee_fixed || 0),
          minAmount: parseFloat(settings.transfer_min_amount || 1),
          maxAmount: parseFloat(settings.transfer_max_amount || 50000),
          dailyLimit: parseFloat(settings.transfer_daily_limit || 100000),
        },
        billboard: {
          enabled: settings.wallet_billboard_enabled === 'true',
          images: settings.wallet_billboard_images
            ? JSON.parse(settings.wallet_billboard_images)
            : [],
          autoPlayInterval: parseInt(settings.wallet_billboard_autoplay_interval || 5000),
        },
      }));
    } catch (error) {
      strapi.log.error('[Wallet] getSettings error:', error);
      ctx.badRequest(utils.errorResponse('WALLET_ERROR', 'Failed to load wallet settings'));
    }
  },
};

File: packages/application/lib/services/wallet.dart
import 'package:chongchana/models/service_response.dart';
import 'package:chongchana/models/wallet.dart';
import 'package:chongchana/models/wallet_transaction.dart';
import 'package:chongchana/models/wallet_settings.dart';
import 'package:chongchana/services/api/fetcher.dart';
import 'package:flutter/material.dart';

class WalletService extends ChangeNotifier {
  Wallet? _wallet;
  List<WalletTransaction> _transactions = [];
  WalletSettings? _settings;
  bool _isLoading = false;
  String? _error;

  Wallet? get wallet => _wallet;
  List<WalletTransaction> get transactions => _transactions;
  WalletSettings? get settings => _settings;
  bool get isLoading => _isLoading;
  String? get error => _error;

  /// Get wallet balance
  Future<Wallet?> getBalance() async {
    try {
      print('[WalletService] getBalance() called');

      // Check if user is authenticated
      final token = await Fetcher.chongjaroenAuth.getAccessToken();
      print('[WalletService] Token present: ${token != null}, Token length: ${token?.length ?? 0}');

      _isLoading = true;
      _error = null;
      notifyListeners();

      final ServiceResponse response = await Fetcher.fetch(
        Fetcher.get,
        '/wallet/balance',
      );

      print('[WalletService] getBalance response - isSuccess: ${response.isSuccess}');
      print('[WalletService] getBalance response - statusCode: ${response.statusCode}');
      print('[WalletService] getBalance response - data: ${response.data}');

      if (response.isSuccess) {
        final data = response.data['data'];
        print('[WalletService] getBalance data: $data');
        _wallet = Wallet(
          userId: data['userId'] ?? 0,
          balance: (data['balance'] is int)
              ? (data['balance'] as int).toDouble()
              : (data['balance'] ?? 0.0).toDouble(),
          pendingBalance: (data['pendingBalance'] is int)
              ? (data['pendingBalance'] as int).toDouble()
              : (data['pendingBalance'] ?? 0.0).toDouble(),
          frozenBalance: (data['frozenBalance'] is int)
              ? (data['frozenBalance'] as int).toDouble()
              : (data['frozenBalance'] ?? 0.0).toDouble(),
          totalBalance: (data['totalBalance'] is int)
              ? (data['totalBalance'] as int).toDouble()
              : (data['totalBalance'] ?? 0.0).toDouble(),
          currency: data['currency'] ?? 'THB',
          status: data['status'] ?? 'active',
          points: data['points'] ?? 0,
          lastTransaction: data['lastTransaction'] != null
              ? DateTime.parse(data['lastTransaction'])
              : null,
        );
        print('[WalletService] Wallet created: balance=${_wallet?.balance}, userId=${_wallet?.userId}');
        _isLoading = false;
        notifyListeners();
        return _wallet;
      } else {
        _error = response.errorMessage ?? 'Failed to fetch balance';
        print('[WalletService] getBalance error: $_error');
        _isLoading = false;
        notifyListeners();
        return null;
      }
    } catch (e) {
      _error = e.toString();
      print('[WalletService] getBalance exception: $e');
      _isLoading = false;
      notifyListeners();
      return null;
    }
  }

  /// Get transaction history
  Future<List<WalletTransaction>> getTransactions({
    int limit = 20,
    int offset = 0,
    String? type,
    String? status,
    String? fromDate,
    String? toDate,
  }) async {
    try {
      print('[WalletService] getTransactions() called');
      _isLoading = true;
      _error = null;
      notifyListeners();

      Map<String, dynamic> params = {
        'limit': limit.toString(),
        'offset': offset.toString(),
      };
      if (type != null) params['type'] = type;
      if (status != null) params['status'] = status;
      if (fromDate != null) params['fromDate'] = fromDate;
      if (toDate != null) params['toDate'] = toDate;

      print('[WalletService] getTransactions params: $params');

      final ServiceResponse response = await Fetcher.fetch(
        Fetcher.get,
        '/wallet/transactions',
        params: params,
      );

      print('[WalletService] getTransactions response - isSuccess: ${response.isSuccess}');
      print('[WalletService] getTransactions response - statusCode: ${response.statusCode}');
      print('[WalletService] getTransactions response - data: ${response.data}');

      if (response.isSuccess) {
        final data = response.data['data'];
        final List<dynamic> transactionsData = data['transactions'] ?? [];
        print('[WalletService] getTransactions found ${transactionsData.length} transactions');

        _transactions = transactionsData.map((tx) {
          return WalletTransaction(
            id: tx['id'] ?? '',
            type: tx['type'] ?? 'unknown',
            amount: (tx['amount'] is int)
                ? (tx['amount'] as int).toDouble()
                : tx['amount'] ?? 0.0,
            balanceAfter: (tx['balanceAfter'] is int)
                ? (tx['balanceAfter'] as int).toDouble()
                : tx['balanceAfter'] ?? 0.0,
            balanceBefore: (tx['balanceBefore'] is int)
                ? (tx['balanceBefore'] as int).toDouble()
                : tx['balanceBefore'] ?? 0.0,
            status: tx['status'] ?? 'unknown',
            paymentMethod: tx['paymentMethod'],
            description: tx['description'],
            referenceId: tx['referenceId'],
            createdAt: tx['createdAt'] != null
                ? DateTime.parse(tx['createdAt'])
                : DateTime.now(),
          );
        }).toList();

        print('[WalletService] Transactions parsed: ${_transactions.length} items');
        _isLoading = false;
        notifyListeners();
        return _transactions;
      } else {
        _error = response.errorMessage ?? 'Failed to fetch transactions';
        print('[WalletService] getTransactions error: $_error');
        _isLoading = false;
        notifyListeners();
        return [];
      }
    } catch (e) {
      _error = e.toString();
      print('[WalletService] getTransactions exception: $e');
      _isLoading = false;
      notifyListeners();
      return [];
    }
  }

  /// Make payment using wallet
  Future<Map<String, dynamic>?> makePayment({
    required double amount,
    required String referenceType,
    required String referenceId,
    String? branchId,
    String? description,
    int usePoints = 0,
  }) async {
    try {
      _isLoading = true;
      _error = null;
      notifyListeners();

      final ServiceResponse response = await Fetcher.fetch(
        Fetcher.post,
        '/wallet/pay',
        params: {
          'amount': amount.toString(),
          'referenceType': referenceType,
          'referenceId': referenceId,
          if (branchId != null) 'branchId': branchId,
          if (description != null) 'description': description,
          'usePoints': usePoints.toString(),
        },
      );

      _isLoading = false;
      notifyListeners();

      if (response.isSuccess) {
        return response.data['data'];
      } else {
        _error = response.errorMessage ?? 'Payment failed';
        return null;
      }
    } catch (e) {
      _error = e.toString();
      _isLoading = false;
      notifyListeners();
      return null;
    }
  }

  /// Redeem voucher
  Future<Map<String, dynamic>?> redeemVoucher(String code) async {
    try {
      _isLoading = true;
      _error = null;
      notifyListeners();

      final ServiceResponse response = await Fetcher.fetch(
        Fetcher.post,
        '/wallet/voucher/redeem',
        params: {'code': code},
      );

      _isLoading = false;
      notifyListeners();

      if (response.isSuccess) {
        return response.data['data'];
      } else {
        _error = response.errorMessage ?? 'Failed to redeem voucher';
        return null;
      }
    } catch (e) {
      _error = e.toString();
      _isLoading = false;
      notifyListeners();
      return null;
    }
  }

  /// Get wallet settings (conversion rates, limits, fees)
  Future<WalletSettings?> getSettings() async {
    try {
      print('[WalletService] getSettings() called');

      final ServiceResponse response = await Fetcher.fetch(
        Fetcher.get,
        '/wallet/settings',
      );

      print('[WalletService] getSettings response - isSuccess: ${response.isSuccess}');
      print('[WalletService] getSettings response - statusCode: ${response.statusCode}');
      print('[WalletService] getSettings response - data: ${response.data}');

      if (response.isSuccess) {
        final data = response.data['data'];
        print('[WalletService] getSettings data: $data');
        _settings = WalletSettings.fromJson(data);
        print('[WalletService] Settings loaded: rate=${_settings?.pointConversion.rate}, minRedemption=${_settings?.pointConversion.minimumRedemption}');
        notifyListeners();
        return _settings;
      } else {
        _error = response.errorMessage ?? 'Failed to fetch settings';
        print('[WalletService] getSettings error: $_error');
        return null;
      }
    } catch (e) {
      _error = e.toString();
      print('[WalletService] getSettings exception: $e');
      return null;
    }
  }

  /// Convert points to wallet credit
  Future<Map<String, dynamic>?> convertPoints(int points) async {
    try {
      _isLoading = true;
      _error = null;
      notifyListeners();

      final ServiceResponse response = await Fetcher.fetch(
        Fetcher.post,
        '/wallet/points/convert',
        params: {'points': points.toString()},
      );

      _isLoading = false;
      notifyListeners();

      if (response.isSuccess) {
        return response.data['data'];
      } else {
        _error = response.errorMessage ?? 'Failed to convert points';
        return null;
      }
    } catch (e) {
      _error = e.toString();
      _isLoading = false;
      notifyListeners();
      return null;
    }
  }

  /// Clear error
  void clearError() {
    _error = null;
    notifyListeners();
  }

  /// Look up user by phone number for transfer
  Future<Map<String, dynamic>?> lookupUserByPhone(String phoneNumber) async {
    try {
      print('[WalletService] lookupUserByPhone() called');
      _isLoading = true;
      _error = null;
      notifyListeners();

      final ServiceResponse response = await Fetcher.fetch(
        Fetcher.post,
        '/wallet/lookup-user',
        params: {'phoneNumber': phoneNumber},
      );

      print('[WalletService] lookupUserByPhone response - isSuccess: ${response.isSuccess}');
      print('[WalletService] lookupUserByPhone response - data: ${response.data}');

      _isLoading = false;
      notifyListeners();

      if (response.isSuccess) {
        return response.data['data'];
      } else {
        _error = response.errorMessage ?? 'Failed to lookup user';
        return null;
      }
    } catch (e) {
      _error = e.toString();
      print('[WalletService] lookupUserByPhone exception: $e');
      _isLoading = false;
      notifyListeners();
      return null;
    }
  }

  /// Transfer funds to another user
  /// Uses idempotency key to prevent double transfers
  Future<Map<String, dynamic>?> transferFunds({
    required int receiverUserId,
    required double amount,
    String? description,
  }) async {
    try {
      print('[WalletService] transferFunds() called');
      print('[WalletService]   receiverUserId: $receiverUserId');
      print('[WalletService]   amount: $amount');

      _isLoading = true;
      _error = null;
      notifyListeners();

      // Generate idempotency key to prevent double transfers
      final idempotencyKey = '${DateTime.now().millisecondsSinceEpoch}_${receiverUserId}_${amount}';

      final ServiceResponse response = await Fetcher.fetch(
        Fetcher.post,
        '/wallet/transfer',
        params: {
          'receiverUserId': receiverUserId.toString(),
          'amount': amount.toString(),
          'description': description ?? '',
          'idempotencyKey': idempotencyKey,
        },
      );

      print('[WalletService] transferFunds response - isSuccess: ${response.isSuccess}');
      print('[WalletService] transferFunds response - data: ${response.data}');

      _isLoading = false;
      notifyListeners();

      if (response.isSuccess) {
        final data = response.data['data'];
        // Refresh wallet balance after successful transfer
        await getBalance();
        return data;
      } else {
        _error = response.errorMessage ?? 'Transfer failed';
        return null;
      }
    } catch (e) {
      _error = e.toString();
      print('[WalletService] transferFunds exception: $e');
      _isLoading = false;
      notifyListeners();
      return null;
    }
  }

  /// Generate payment QR code for in-store payments
  /// Returns JWT token to be encoded in QR code
  Future<Map<String, dynamic>?> generatePaymentQR({
    String purpose = 'store_payment', // or 'beer_machine'
  }) async {
    try {
      print('[WalletService] generatePaymentQR() called');
      print('[WalletService]   purpose: $purpose');

      _isLoading = true;
      _error = null;
      notifyListeners();

      final ServiceResponse response = await Fetcher.fetch(
        Fetcher.post,
        '/wallet/payment-qr/generate',
        params: {
          'purpose': purpose,
        },
      );

      print('[WalletService] generatePaymentQR response - isSuccess: ${response.isSuccess}');
      print('[WalletService] generatePaymentQR response - data: ${response.data}');

      _isLoading = false;
      notifyListeners();

      if (response.isSuccess) {
        return response.data['data'];
      } else {
        _error = response.errorMessage ?? 'Failed to generate payment QR';
        return null;
      }
    } catch (e) {
      _error = e.toString();
      print('[WalletService] generatePaymentQR exception: $e');
      _isLoading = false;
      notifyListeners();
      return null;
    }
  }

  /// Get payment QR history
  Future<List<dynamic>> getPaymentQRHistory({
    int page = 1,
    int limit = 20,
    String? type, // 'store_payment' or 'beer_machine_payment'
  }) async {
    try {
      print('[WalletService] getPaymentQRHistory() called');
      _isLoading = true;
      _error = null;
      notifyListeners();

      Map<String, dynamic> params = {
        'page': page.toString(),
        'limit': limit.toString(),
      };
      if (type != null) params['type'] = type;

      final ServiceResponse response = await Fetcher.fetch(
        Fetcher.get,
        '/wallet/payment-qr/history',
        params: params,
      );

      print('[WalletService] getPaymentQRHistory response - isSuccess: ${response.isSuccess}');
      print('[WalletService] getPaymentQRHistory response - data: ${response.data}');

      _isLoading = false;
      notifyListeners();

      if (response.isSuccess) {
        final data = response.data['data'];
        return data['transactions'] ?? [];
      } else {
        _error = response.errorMessage ?? 'Failed to fetch payment history';
        return [];
      }
    } catch (e) {
      _error = e.toString();
      print('[WalletService] getPaymentQRHistory exception: $e');
      _isLoading = false;
      notifyListeners();
      return [];
    }
  }

  /// Clear all wallet data (called on logout)
  void clearWalletData() {
    _wallet = null;
    _transactions = [];
    _settings = null;
    _isLoading = false;
    _error = null;
    notifyListeners();
  }
}
